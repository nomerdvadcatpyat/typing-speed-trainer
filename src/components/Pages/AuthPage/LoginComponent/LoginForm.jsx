import React from 'react';
import {Link} from "react-router-dom";
import '../AuthForm.scss'
import {AuthInput} from "../AuthInput/AuthInput";
import PropTypes from "prop-types";
import {Button, Form} from "react-bootstrap";


export function LoginForm(props) {
	const {
		handleSubmit,
		handleChange,
		handleBlur,
		touched,
		errors,
		values,
		status,
		isSubmitting
	} = props;

	return (
			<Form className="auth-form login" onSubmit={handleSubmit}>
				<AuthInput
					className={`auth-form__input ${touched.login && errors.login ? 'invalid' : ''}`}
					name="login"
					type="text"
					placeholder="Login"
					errors={touched.login && errors.login}
					onChange={handleChange}
					onBlur={handleBlur}
					value={values.login}
				/>

				<AuthInput
					className={`auth-form__input ${touched.password && errors.password ? 'invalid' : ''}`}
					name="password"
					type="password"
					placeholder="Password"
					errors={touched.password && errors.password}
					onChange={handleChange}
					onBlur={handleBlur}
					value={values.password}
				/>

				{
					errors.all ? (
						<span className="auth-form__error-message"> {errors.all} </span>
					) : null
				}

				<Button
					className="auth-form__submit"
					type="submit"
					variant="secondary"
					disabled={
						values.login.length === 0 ||
						values.password.length === 0 ||
						(status && !status.success) ||
						isSubmitting
					}>
					Submit
				</Button>

				<Link to="/auth/registration" className="auth-form__link-to-register"> Нет аккаунта? Зарегистрироваться </Link>
			</Form>
	);
}


LoginForm.propTypes = {
	onSubmit: PropTypes.func,
	formError: PropTypes.string
}