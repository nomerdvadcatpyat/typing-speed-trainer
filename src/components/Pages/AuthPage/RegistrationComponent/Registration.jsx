import React, {useState} from 'react';
import {ControlledInput} from "../../../ControlledInput/ControlledInput";
import '../AuthForm.scss'

export const Registration = ({ registration }) => {
	const [email, setEmail] = useState("");
	const [password, setPassword] = useState("");
	const [rePassword, setRePassword] = useState("");

	return (
		<div className="auth-form registration">

			<div className="input-field">
				<i className="material-icons prefix">email</i>
				<ControlledInput
					id="register-email-input"
					type="text"
					value={email}
					onChange={setEmail}
				/>
				<label htmlFor="register-email-input">Email</label>
			</div>

			<div className="input-field">
				<i className="material-icons prefix">lock</i>
				<ControlledInput
					id="register-password-input"
					type="password"
					value={password}
					onChange={setPassword}
				/>
				<label htmlFor="register-password-input">Password</label>
			</div>

			<div className="input-field">
				<i className="material-icons prefix">lock_outline</i>
				<ControlledInput
					id="register-re-password-input"
					type="password"
					value={rePassword}
					onChange={setRePassword}
				/>
				<label htmlFor="register-re-password-input">Confirm password</label>
			</div>

			<button
				className="btn waves-effect"
				onClick={() => registration(email, password, rePassword)}
			>
				Submit
			</button>

		</div>
	);
}