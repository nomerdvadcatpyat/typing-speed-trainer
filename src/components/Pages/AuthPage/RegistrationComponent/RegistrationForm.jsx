import React from 'react';
import '../AuthForm.scss'
import {AuthInput} from "../AuthInput/AuthInput";
import {Button} from "react-bootstrap";


export const RegistrationForm = (props) => {
	const {
		handleSubmit,
		handleChange,
		handleBlur,
		touched,
		errors,
		values,
		status,
		isSubmitting
	} = props;

	return (
			<form className="auth-form registration" onSubmit={handleSubmit}>
				<AuthInput
					className={`auth-form__input ${touched.login && errors.login ? 'invalid' : ''}`}
					name="login"
					type="text"
					placeholder="Login"
					errors={touched.login && errors.login}
					onChange={handleChange}
					onBlur={handleBlur}
					value={values.login}
				/>

				<AuthInput
					className={`auth-form__input ${touched.password && errors.password ? 'invalid' : ''}`}
					name="password"
					type="password"
					placeholder="Password"
					errors={touched.password && errors.password}
					onChange={handleChange}
					onBlur={handleBlur}
					value={values.password}
				/>

				<AuthInput
					className={`auth-form__input ${touched.rePassword && errors.rePassword ? 'invalid' : ''}`}
					name="rePassword"
					type="password"
					placeholder="Confirm password"
					errors={touched.rePassword && errors.rePassword}
					onChange={handleChange}
					onBlur={handleBlur}
					value={values.rePassword}
				/>

				{
					errors.all ? (
						<span className="auth-form__error-message"> {errors.all} </span>
					) : null
				}

				<Button
					className="auth-form__submit"
					variant="dark"
					disabled={
						values.login.length === 0 ||
						values.password.length === 0 ||
						values.rePassword.length === 0 ||
						(status && !status.success) ||
						isSubmitting
					}
					type="submit"
				>
					Submit
				</Button>
			</form>
	);
}
