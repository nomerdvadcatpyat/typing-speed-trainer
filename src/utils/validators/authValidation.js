import * as yup from "yup";

export const registrationSchema = yup.object().shape({
	login: yup.string().min(2).max(25).required("Login is required"),
	password: yup.string().min(3).max(25).required('Password is required'),
	rePassword: yup.string().min(3).max(25)
		.oneOf([yup.ref('password'), null], 'Passwords must match')
});

export const loginSchema = yup.object().shape({
	login: yup.string().min(2).max(25).required('Login is required'),
	password: yup.string().min(3).max(25).required('Password is required')
});

