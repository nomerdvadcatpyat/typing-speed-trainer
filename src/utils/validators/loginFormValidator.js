
export const requiredField = (value) => {
	if(!value) 	return 'Поле обязательно для заполнения';

	return undefined;
}

export const maxLengthCreator = maxLength => value => {
	if(value.length > maxLength) return `Поле не должно быть больше ${maxLength} символов`;

	return undefined;
}


