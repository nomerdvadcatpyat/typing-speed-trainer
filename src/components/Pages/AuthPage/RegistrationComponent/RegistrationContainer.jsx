import React from 'react';
import {connect} from 'react-redux';
import {Registration} from "./Registration";
import {registration} from "../../../../store/actions/auth";
import {bindActionCreators} from "redux";

function mapDispatchToProps(dispatch) {
	return {
		registration: bindActionCreators(registration, dispatch)
	};
}

const RegistrationContainer = (props) => {
	return (
		<Registration {...props} />
	);
}

export default connect(
	null,
	mapDispatchToProps
)(RegistrationContainer);