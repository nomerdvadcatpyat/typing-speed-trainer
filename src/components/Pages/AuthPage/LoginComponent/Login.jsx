import React, {useState} from 'react';
import {Link} from "react-router-dom";
import '../AuthForm.scss'
import {loginSchema} from "../../../../utils/validators/authValidation";


export function LoginForm(props) {

	const [email, setEmail] = useState('');
	const [password, setPassword] = useState('');

	const submitHandler = async (e) => {
		console.log('onSubmit', email, password)
		e.preventDefault();

	  const isValid = await loginSchema.isValid({email, password})

		console.log(isValid);

		if(isValid)
			props.tryLogin({email, password});
	}

	return (
		<form onSubmit={submitHandler} className="auth-form login">

			<div className="input-field form-control">
				<i className="material-icons prefix">email</i>
				<input
					id="login-email-input"
					name="email"
					onChange={e => setEmail(e.target.value)}
					value={email}
					type="text"
				/>
				<label htmlFor="login-email-input">Email</label>
			</div>

			<div className="input-field form-control">
				<i className="material-icons prefix">lock_outline</i>
				<input
					id="login-password-input"
					name="password"
					onChange={e => setPassword(e.target.value)}
					value={password}
					type="password"
				/>
				<label htmlFor="login-password-input">Password</label>
			</div>


			<button className="btn waves-effect">
				Submit
			</button>

			<Link to="/auth/registration" className="link-to-register"> Нет аккаунта? Зарегистрироваться </Link>

		</form>
	);
}


// export const LoginReduxForm = ({ tryLogin }) => {
// 	const handleSubmit = (inputUser) => {
// 		console.log(inputUser);
// 		tryLogin(inputUser);
// 	}
//
// 	return withReduxForm(LoginForm, 'login', handleSubmit)
// }
