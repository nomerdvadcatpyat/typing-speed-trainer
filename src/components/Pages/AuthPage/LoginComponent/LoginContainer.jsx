import React from 'react';
import {connect} from 'react-redux';
import {bindActionCreators, compose} from "redux";
import {setUserActionCreator} from "../../../../store/actionCreators/userActionCreators";
import {LoginForm} from "./LoginForm";
import {login} from "../../../../utils/api/authApi";
import {Redirect} from "react-router-dom";
import PropTypes from "prop-types";
import {withFormik} from "formik";
import {loginSchema} from "../../../../utils/validators/authValidation";


const LoginContainer = (props) => {
	return props.status && props.status.success ? (<Redirect to="/" />) :
				(<LoginForm {...props} />)
}


const mapDispatchToProps = (dispatch) => {
	return {
		setUser: bindActionCreators(setUserActionCreator, dispatch)
	};
}

async function handleSubmit(values, { props, setErrors, setStatus }) {
	try {
		const data = await login(values);
		setStatus({ success: true});
		props.setUser(data);
	} catch (e) {
		setErrors({ all: e.response.data.error });
	}
}

export default compose(
	connect(null, mapDispatchToProps),
	withFormik({
		mapPropsToValues: () => ({
			email: "",
			password: ""
		}),
		handleSubmit,
		validationSchema: loginSchema
	})
)(LoginContainer);


LoginContainer.propTypes = {
	setUser: PropTypes.func
}