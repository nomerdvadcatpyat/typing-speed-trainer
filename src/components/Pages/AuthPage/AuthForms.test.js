import {LoginForm} from "./LoginComponent/LoginForm";
import {mount} from "enzyme";
import {loginSchema} from "../../../utils/validators/authValidation";
import {FormikProvider, useFormik} from "formik";
import {BrowserRouter} from "react-router-dom";
import {RegistrationForm} from "./RegistrationComponent/RegistrationForm";


let handleSubmit;
beforeEach(() => {
	handleSubmit = jest.fn();
});


const FormWithFormik = ({InnerForm, values}) => {
	const formik = useFormik({
		initialValues: values,
		onSubmit: handleSubmit,
		validationSchema: loginSchema
	});

	return (
		<FormikProvider value={formik}>
			<BrowserRouter>
				<InnerForm
					handleSubmit={handleSubmit}
					handleChange={formik.handleChange}
					handleBlur={formik.handleBlur}
					touched={formik.touched}
					errors={formik.errors}
					values={values}
					status={formik.status}
					isSubmitting={formik.isSubmitting}
				/>
			</BrowserRouter>
		</FormikProvider>
	);
}



const correctLoginValues = {
	email: "test@test.com",
	password: "111"
}

const incorrectLoginEmail = {
	email: "testest.com",
	password: "111"
}

describe('<LoginForm />', () => {

	it('should be mounted', () => {
		const loginForm = mount(<FormWithFormik InnerForm={LoginForm} values={correctLoginValues}/>);

		expect(loginForm).toMatchSnapshot();
	});


	it('submit button should be active', () => {
		const loginForm = mount(<FormWithFormik InnerForm={LoginForm} values={correctLoginValues}/>);

		expect(loginForm.find('.auth-form__submit[disabled=""]').length).toBe(0);
	});


	it('form should be submitted', () => {
		const loginForm = mount(<FormWithFormik InnerForm={LoginForm} values={correctLoginValues}/>);
		loginForm.find('.auth-form').simulate('submit');

		expect(handleSubmit).toHaveBeenCalledTimes(1);
	});


	it('submit button should be disabled', () => {
		const loginForm = mount(<FormWithFormik InnerForm={LoginForm} values={incorrectLoginEmail}/>);

		expect(loginForm.find('.auth-form__submit[disabled]').length).toBe(1);
	});

});




const correctRegistrationValues = {
	email: "test@test.com",
	password: "111",
	rePassword: "111"
}

describe('<RegistrationForm />', () => {

	it('should be mounted', () => {
		const registerForm = mount(<FormWithFormik InnerForm={RegistrationForm} values={correctRegistrationValues}/>);

		expect(registerForm).toMatchSnapshot();
	});


	it('submit button should be active', () => {
		const registerForm = mount(<FormWithFormik InnerForm={RegistrationForm} values={correctRegistrationValues}/>);

		expect(registerForm.find('.auth-form__submit[disabled=""]').length).toBe(0);
	});


	it('form should be submitted', () => {
		const registerForm = mount(<FormWithFormik InnerForm={RegistrationForm} values={correctRegistrationValues}/>);
		registerForm.find('.auth-form').simulate('submit');

		expect(handleSubmit).toHaveBeenCalledTimes(1);
	});


	it('submit button should be disabled. Incorrect Email.', () => {
		const incorrectRegistrationEmail = {
			email: "testest.com",
			password: "111",
			rePassword: "111"
		}
		const registerForm = mount(<FormWithFormik InnerForm={RegistrationForm} values={incorrectRegistrationEmail}/>);

		expect(registerForm.find('.auth-form__submit[disabled]').length).toBe(1);
	});

	it('submit button should be disabled. Incorrect Password Confirmation.', () => {
		const incorrectRegistrationPasswordConfirmation = {
			email: "testest@asd.com",
			password: "111",
			rePassword: "11321"
		}
		const registerForm = mount(<FormWithFormik InnerForm={RegistrationForm}
		                                        values={incorrectRegistrationPasswordConfirmation}/>);

		expect(registerForm.find('.auth-form__submit[disabled]').length).toBe(1);
	});

	it('submit button should be disabled. Too short password.', () => {
		const incorrectRegistrationShortPassword = {
			email: "testest@asd.com",
			password: "1",
			rePassword: "1"
		}
		const registerForm = mount(<FormWithFormik InnerForm={RegistrationForm}
		                                        values={incorrectRegistrationShortPassword}/>);

		expect(registerForm.find('.auth-form__submit[disabled]').length).toBe(1);
	});

});






