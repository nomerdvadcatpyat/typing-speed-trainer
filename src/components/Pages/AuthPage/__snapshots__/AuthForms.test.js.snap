// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> should be mounted 1`] = `
<FormWithFormik
  InnerForm={[Function]}
  values={
    Object {
      "email": "test@test.com",
      "password": "111",
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "typingState": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <LoginForm
        errors={Object {}}
        handleBlur={[Function]}
        handleChange={[Function]}
        handleSubmit={[MockFunction]}
        isSubmitting={false}
        touched={Object {}}
        values={
          Object {
            "email": "test@test.com",
            "password": "111",
          }
        }
      >
        <form
          className="auth-form login"
          onSubmit={[MockFunction]}
        >
          <AuthInput
            className=""
            icon="email"
            id="login-email-input"
            labelText="Email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="test@test.com"
          >
            <div
              className="auth-input input-field"
            >
              <i
                className="auth-input__icon small material-icons"
              >
                email
              </i>
              <Field
                className=""
                id="login-email-input"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value="test@test.com"
              >
                <input
                  className=""
                  id="login-email-input"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="test@test.com"
                />
              </Field>
              <label
                htmlFor="login-email-input"
              >
                Email
              </label>
            </div>
          </AuthInput>
          <AuthInput
            className=""
            icon="lock"
            id="login-password-input"
            labelText="Password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value="111"
          >
            <div
              className="auth-input input-field"
            >
              <i
                className="auth-input__icon small material-icons"
              >
                lock
              </i>
              <Field
                className=""
                id="login-password-input"
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                type="password"
                value="111"
              >
                <input
                  className=""
                  id="login-password-input"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="password"
                  value="111"
                />
              </Field>
              <label
                htmlFor="login-password-input"
              >
                Password
              </label>
            </div>
          </AuthInput>
          <button
            className="btn waves-effect auth-form__submit"
            disabled={false}
            type="submit"
          >
            Submit
          </button>
          <Link
            className="auth-form__link-to-register"
            to="/auth/registration"
          >
            <LinkAnchor
              className="auth-form__link-to-register"
              href="/auth/registration"
              navigate={[Function]}
            >
              <a
                className="auth-form__link-to-register"
                href="/auth/registration"
                onClick={[Function]}
              >
                 Нет аккаунта? Зарегистрироваться 
              </a>
            </LinkAnchor>
          </Link>
        </form>
      </LoginForm>
    </Router>
  </BrowserRouter>
</FormWithFormik>
`;

exports[`<RegistrationForm /> should be mounted 1`] = `
<FormWithFormik
  InnerForm={[Function]}
  values={
    Object {
      "email": "test@test.com",
      "password": "111",
      "rePassword": "111",
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "typingState": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <RegistrationForm
        errors={Object {}}
        handleBlur={[Function]}
        handleChange={[Function]}
        handleSubmit={[MockFunction]}
        isSubmitting={false}
        touched={Object {}}
        values={
          Object {
            "email": "test@test.com",
            "password": "111",
            "rePassword": "111",
          }
        }
      >
        <form
          className="auth-form registration"
          onSubmit={[MockFunction]}
        >
          <AuthInput
            className=""
            icon="email"
            id="register-email-input"
            labelText="Email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="test@test.com"
          >
            <div
              className="auth-input input-field"
            >
              <i
                className="auth-input__icon small material-icons"
              >
                email
              </i>
              <Field
                className=""
                id="register-email-input"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value="test@test.com"
              >
                <input
                  className=""
                  id="register-email-input"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="text"
                  value="test@test.com"
                />
              </Field>
              <label
                htmlFor="register-email-input"
              >
                Email
              </label>
            </div>
          </AuthInput>
          <AuthInput
            className=""
            icon="lock"
            id="register-password-input"
            labelText="Password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value="111"
          >
            <div
              className="auth-input input-field"
            >
              <i
                className="auth-input__icon small material-icons"
              >
                lock
              </i>
              <Field
                className=""
                id="register-password-input"
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                type="password"
                value="111"
              >
                <input
                  className=""
                  id="register-password-input"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="password"
                  value="111"
                />
              </Field>
              <label
                htmlFor="register-password-input"
              >
                Password
              </label>
            </div>
          </AuthInput>
          <AuthInput
            className=""
            icon="lock_outline"
            id="register-re-password-input"
            labelText="Confirm password"
            name="rePassword"
            onBlur={[Function]}
            onChange={[Function]}
            type="password"
            value="111"
          >
            <div
              className="auth-input input-field"
            >
              <i
                className="auth-input__icon small material-icons"
              >
                lock_outline
              </i>
              <Field
                className=""
                id="register-re-password-input"
                name="rePassword"
                onBlur={[Function]}
                onChange={[Function]}
                type="password"
                value="111"
              >
                <input
                  className=""
                  id="register-re-password-input"
                  name="rePassword"
                  onBlur={[Function]}
                  onChange={[Function]}
                  type="password"
                  value="111"
                />
              </Field>
              <label
                htmlFor="register-re-password-input"
              >
                Confirm password
              </label>
            </div>
          </AuthInput>
          <button
            className="btn waves-effect auth-form__submit"
            disabled={false}
            type="submit"
          >
            Submit
          </button>
        </form>
      </RegistrationForm>
    </Router>
  </BrowserRouter>
</FormWithFormik>
`;
